`str length`, `str substring`, `str index-of`, `split words` and `split chars` now use graphemes instead of UTF-8 bytes

webbedspace opened the pull request:
# Description

EDIT: Now includes `split chars`! `split chars` formerly used code points (I assume).

Closes #7742.

BEFORE:
```
〉'🇯🇵ほげ ふが ぴよ'   | str substring 4..6
�
〉'🇯🇵ほげ ふが ぴよ'   | str length
28
〉'🇯🇵ほげ ふが ぴよ'   | str index-of 'ふが'
15
〉'🇯🇵ほげ' | split chars | to nuon
[🇯, 🇵, ほ, げ]
```

AFTER:
```
〉'🇯🇵ほげ ふが ぴよ'   | str substring 4..6 -g
ふが
〉'🇯🇵ほげ ふが ぴよ'   | str length -g
9
〉'🇯🇵ほげ ふが ぴよ'   | str index-of 'ふが' -g
4
〉'🇯🇵ほげ' | split chars -g | to nuon
[🇯🇵, ほ, げ]
```

The previous behaviour is so obviously wrong that I feel comfortable not even providing a flag for enabling the old behaviour.

Extra reading: https://manishearth.github.io/blog/2017/01/14/stop-ascribing-meaning-to-unicode-code-points/

>Basically, unless you have very specific requirements or are able to query the font, use an API that segments strings into grapheme clusters wherever you need to deal with the notion of “character”.

# User-Facing Changes

See above. Note this has almost no impact on text which only contains ASCII. With one exception: `"\r\n"` is now considered to have a length of 1.

# Tests + Formatting

Don't forget to add tests that cover your changes.

Make sure you've run and fixed any issues with these commands:

- `cargo fmt --all -- --check` to check standard code formatting (`cargo fmt --all` applies these changes)
- `cargo clippy --workspace -- -D warnings -D clippy::unwrap_used -A clippy::needless_collect` to check that you're using the standard code style
- `cargo test --workspace` to check that all tests pass

# After Submitting

If your PR had any user-facing changes, update [the documentation](https://github.com/nushell/nushell.github.io) after the PR is merged, if necessary. This will help us keep the docs up to date.

fdncred commented:
I'm not sure how I feel about this. I come from c/c++ which, I believe, returns bytes from str length. c# returns the number of Chars vs unicode characters. Maybe it's just my background that makes this seem foreign or wrong.

Personally, I'd rather have a way to also return bytes? Maybe the bytes commands can/could do that? But I'm willing to go with the core-team's decision. So, I'll wait to see what input they offer.

jntrnr commented:
My two cents while we try to figure out what the right call is:

I would like this if we can figure out a way to *always* treat strings this way. Meaning that if people work with strings and ranges, they know the ranges will always be over graphemes. Same for indexing.

@fdncred agreed about always having a way to get to the bytes. As a workaround, we could ask folks to use `into binary` as an intermediate step to make it clear they want bytes.

The trouble would be if sometimes strings work one way and sometimes they work another, they people are going to get confused. For example, while it's not yet supported, we might want to add cell paths on strings in the future to make it easier to get to a substring. If we do, and take the grapheme approach, then the cell path would need to do the same. (just as one example)

rgwood commented:
^agreed.

I'm up for landing this and encouraging people to use `into binary` if they want the raw bytes.

fdncred commented:
If we do this, we need to take a look at #7757.  We may also need to look at `split chars`, I haven't looked to see if it works on bytes or code points. e.g. `good nushell' | split chars | range 5..12 | str join`.

webbedspace commented:
Word-splitting is already provided by [unicode_segmentation](https://docs.rs/unicode-segmentation/latest/unicode_segmentation/struct.UnicodeWords.html), isn't it?

fdncred commented:
> Word-splitting is already provided by [unicode_segmentation](https://docs.rs/unicode-segmentation/latest/unicode_segmentation/struct.UnicodeWords.html), isn't it?

I don't believe so. If you modify that one two, please make sure the commented out examples, at the end of the file, still return identical results. I created it to work a specific way and I'd really like to keep it that way if possible. 

I still think this PR is questionable because it really seems like the entire code base needs to be gone through looking for these type of things that you're pointing out and changing. As an example, do we need to change everywhere we use chars() or split()? Above all, I think we want to be consistent with handling these type of things.

bobhy commented:
The grapheme-based results are clearly better for the Nushell programming model, that's what most people are thinking about when counting "characters".

However it will be perpetually confusing if we try to make a `str` command deliver these results.  I don't think Rust or C++ will ever change their `str`.  So, I propose the grapheme-based results all be cloned into a new command perhaps called `grapheme` or `gme` that has all the same subcommands as `str`.

I'd extend the notion of grapheme to include ANSI escape sequences.  See #7757.

rgwood commented:
> However it will be perpetually confusing if we try to make a `str` command deliver these results.

I disagree. I think most Nu users will expect `str` commands to use graphemes instead of bytes.

webbedspace commented:
>I don't believe so. If you modify that one too, please make sure the commented out examples, at the end of the file, still return identical results. I created it to work a specific way and I'd really like to keep it that way if possible.

So wait, first I find out I can't run `cargo test` cos I need `cargo test --workspace`, *now* you're tellin' me I gotta run tests that aren't even in the code? What kinda codebase is this? 😼

Anyway, I just checked and it outputs the same results. I couldn't benchmark it because `benchmark` apparently doesn't freaking work with variables:

```
〉let alice = (curl -sL "https://www.gutenberg.org/files/11/11-0.txt")
〉benchmark { $alice | str downcase | split words | uniq -c | sort-by count --reverse | first 10 }
Error: nu::shell::variable_not_found (link)

  × Variable not found
   ╭─[entry #12:1:1]
 1 │ benchmark { $alice | str downcase | split words | uniq -c | sort-by count --reverse | first 10 }
   ·             ───┬──
   ·                ╰── variable not found
   ╰────
```

fdncred commented:
> Anyway, I just checked and it outputs the same results (…after the most recent commit 🐟). I couldn't benchmark it because benchmark apparently doesn't freaking work with variables:

I'm glad the results are the same. Performance is important too. One way to make a test out of this is to put the file at https://www.gutenberg.org/files/11/11-0.txt in our test data.

A little background on `split words`. I wrote it specifically to be able to this but easier. https://datascienceatthecommandline.com/2e/chapter-4-creating-command-line-tools.html#converting-one-liners-into-shell-scripts. I was thinking that much of the examples in this book could be done easier/more intuitively than with existing tools and this was a good example/place to start.

This:
```
> curl -sL "https://www.gutenberg.org/files/11/11-0.txt" | 
tr '[:upper:]' '[:lower:]' | 
grep -oE "[a-z\']{2,}" | 
sort | 
uniq -c | 
sort -nr | 
head -n 10 

   1839 the
    942 and
    811 to
    638 of
    610 it
    553 she
    486 you
    462 said
    435 in
    403 alice
```
vs This:
```
> fetch https://www.gutenberg.org/files/11/11-0.txt | 
str downcase | 
split words -l 2 | 
uniq -c | 
sort-by count --reverse | 
first 10

╭───┬───────┬───────╮
│ # │ value │ count │
├───┼───────┼───────┤
│ 0 │ the   │  1839 │
│ 1 │ and   │   942 │
│ 2 │ to    │   811 │
│ 3 │ of    │   638 │
│ 4 │ it    │   610 │
│ 5 │ she   │   553 │
│ 6 │ you   │   486 │
│ 7 │ said  │   462 │
│ 8 │ in    │   435 │
│ 9 │ alice │   403 │
╰───┴───────┴───────╯
```

sholderbach commented:
While I absolutely see the value of abstracting the complexity of encoding by operating on *graphemes* I think there are some issues if we were to commit to that from the perspective of the use as a programming language.

One clear benefit of the grapheme definition would be that in the interactive use you would count "characters" manually for most scripts (as in language alphabets) on the grapheme boundaries.

The two downsides I see:

- Graphemes have some surprising properties for things that are kind of common: `CRLF`/`\r\n` is one grapheme even though programmers expect to sometimes deal with it and remember that it is two ASCII `char`s or two codepoints/scalars in UTF8/
- Working with graphemes (and to a slightly lesser extent codepoints*) turns index operations that were previously `O(1)` into a `O(n)` operation. On small strings that is probably neglible and just a small factor, but in the general programming usecase it could turn an otherwise linear operation into something `O(n^2)` that would scale poorly on larger datasets.

> *) for codepoints you don't need an extra lookup as the length of a codepoint is part of the UTF8 encoding and thus purely something happening on the instruction level.

bobhy commented:
So, graphemes (or is it grapheme clusters?) are so much more what all the
non-english-users want to be dealing with that we should have really solid
reasons for imposing that handicap on them.
Many data streams represent newline as a single character natively, that's
a familiar option to most folk already.
And while indexing is certainly slower for grapheme (and for utf8 strings),
*Iteration* is the same O(n) operation for bytes and graphemes. Parsing and
lexing depend on iteration, they are common use cases.  What kinds of apps
are we handicapping by slowing down direct indexing?

On Sun, Jan 15, 2023, 4:51 PM Stefan Holderbach ***@***.***>
wrote:

> While I absolutely see the value of abstracting the complexity of encoding
> by operating on *graphemes* I think there are some issues if we were to
> commit to that from the perspective of the use as a programming language.
>
> One clear benefit of the grapheme definition would be that in the
> interactive use you would count "characters" manually for most scripts (as
> in language alphabets) on the grapheme boundaries.
>
> The two downsides I see:
>
>    - Graphemes have some surprising properties for things that are kind
>    of common: CRLF/\r\n is one grapheme even though programmers expect to
>    sometimes deal with it and remember that it is two ASCII chars or two
>    codepoints/scalars in UTF8/
>    - Working with graphemes (and to a slightly lesser extent codepoints*)
>    turns index operations that were previously O(1) into a O(n)
>    operation. On small strings that is probably neglible and just a small
>    factor, but in the general programming usecase it could turn an otherwise
>    linear operation into something O(n^2) that would scale poorly on
>    larger datasets.
>
> *) for codepoints you don't need an extra lookup as the length of a
> codepoint is part of the UTF8 encoding and thus purely something happening
> on the instruction level.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nushell/nushell/pull/7752#issuecomment-1383261975>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPBIZFCIIHB35L33J6AZG3WSRWPDANCNFSM6AAAAAAT3GQ6NI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

webbedspace commented:
>So, graphemes (or is it grapheme clusters?)

The actual correct name for these is "extended grapheme clusters" ("extended" because this is a "fixed" version of the original grapheme cluster specification, which I opt into by calling `.graphemes(true)`). That being said, since there is no real correspondence between UTF8 bytes, code points, or even single graphemes, referring to clusters as "graphemes" for short (which is what unicode_segmentation does) doesn't necessarily incorrectly label one thing as another.

bobhy commented:
(thanks for useful clarification on grapheme terminology, helpful for the naive english-speaker)
My takeaway from this whole conversation is that I want a different view of strings, as being composed of  'print positions', entities that take up visible space when you print them out (or don't).  And I think the way to get that is to provide a whole new command analogous to `str` that does length arithmetic that's closer to what I see when I print it out.  
I'm reading that Unicode has multi grapheme sequences such as composing dieresis that will still complicate indexing arithmetic even if we accept this PR and fix `str length` for Unicode.   I want my 'print position' arithmetic to also work when strings have ANSI color escape sequences, that  has nothing to do with Unicode.  
So I  won't muddy the waters further by offering an uninformed and inexpert opinion.  If it's put to a vote however, I'll vote to accept the PR, it does seem to improve the consistency of (Nushell) str arithmetic.

fdncred commented:
I think I'm voting for `--grapheme(-g)` parameters on these commands so we can do bytes and graphemes if directed, defaulting to bytes.

webbedspace commented:
I'll only begrudgingly agree to a non-default flag if an additional config option is added that makes graphemes the default instead (and has another flag that flips it back to bytes, similar to `rm` and `always_trash`).

Also, reminder that `split chars` never used bytes and instead uses code points, so if you want things to "default to bytes" (which means UTF8 bytes) then you'll have to agree to changing that.

sholderbach commented:
In today's meeting we concluded that we don't feel fully comfortable to make a radical breaking change to the core data type of `String` (yet). We don't fully know the blast radius of code in the wild that depends on the current indexing semantics.

Nushell should definetely support you in interacting  with the complexity of international languages and other Unicode symbols.
As Manish pointed out in the end of the blog post, for interaction with Graphemes you typically want to do iteration in one fell swoop. So providing an iterator command that emits a `PipelineData::ListStream` of grapheme strings would be great. (e.g. `split grapheme` to accompany `split chars` or flag to not break compatibility at first). But dealing with the potential fall out to making a change to indexing feels risky.

For the indexing/index-emitting commands we would like to have that specialized through either a dedicated command name or flag for the moment.
Doing the breaking change just yet didn't get the confidence of the collected team. 

In a low level language indexing through byte *slices* is great for performance but for higher level operations we want to clarify good semantics that both support hacking (e.g. python having `\r\n` as two separate chars through codepoint level) and natural language ops (GO graphemes)

During the meeting we also discussed that we probably need to become more encoding aware in the future. While on the one hand we should move into the well defined well abstracted world of graphemes, the messy real world still has files or file paths that are encoded with non UTF-8 or even non Unicode things. 
While some of this is part of how we deal with the detailed implementation (and being forced not to use `String/&str` in some places) we might want to also consider what the semantics are for nushell users that want to work with old encodings. Do we want to expect to provide them byte level access or do we keep separate semantics.
(Came up as part of the idea to maybe have a `"Hällö Ümláüt" | decode graphemes | ... some ops | ` like command to move the segmentation decision to a type system level)

webbedspace commented:
I've now added flags to the affected commands. Since I sincerely believe having to opt into the grapheme flag hundreds of times is going to be a major pain in the patoot, I've also added a config flag (as I offered to do earlier) that looks like this:
```
str: {
  grapheme_clusters: false
}
```
This config flag is only used when both of the new flags are absent.

The reason I created a "str" record in $env.config is because I can easily anticipate other `str` config options being added in future (such as locale-based sorting or locale-based capitalisation).

So, the new flags are:
* `str` subcommands: --utf-8-bytes(-b), --graphemes(-g)
* `split chars`, `split words`: --code-points(-c), --graphemes(-g) ←these commands use code points instead of UTF-8 bytes, which is NOT consistent with `str`'s commands.

sholderbach commented:
A configuration option to affect core semantics of code that might be commonly used in scripts and not just some interactive detail just opens a whole new can of worms. I don't like that.

webbedspace commented:
There's already `float_precision` as well as `ENV_CONVERSIONS`, so it's not unprecedented.

webbedspace commented:
OK, look. I want to put it out here so that everyone is aware that making the human-centered option non-default, with no way to force it default, is just bad UX design, and I don't support it. That being said, I have now discarded the aformentioned config option for now. I'm not removing the flags (-b and -c) that "restore" the default behaviour because having them available can allow graphemes to be switched to default later, while providing code with a future-proofed and explicit way of asking for UTF8 byte lengths (as I said above).

fdncred commented on Cargo.toml:
```diff
@@ -31,7 +31,7 @@ members = [
 	"crates/nu-plugin",
 	"crates/nu_plugin_inc",
-	"crates/nu_plugin_gstat",
+	#"crates/nu_plugin_gstat",
```
I'd prefer not to lose the gstat plugin.

webbedspace replied:
Sorry, that was a testing thing I had to do to get `cargo test --workspace` to run.

fdncred commented:
Hey @webbedspace, We really appreciate all your support and work on this and other PRs. We're not trying to make bad design decisions nor are we trying to upset you. TBH, we're still discussing exactly what this should look like.

One thought is to do exactly what you've done here and what I personally voted for above, which is adding parameters. However, there's another school of thought that we may not want to have parameters. We may want to have an entire separate set of commands that deal with graphemes. Those would be some commands that are somewhat analogous to the `str` commands and could also include `decode grapheme` and `into grapheme`.

What are your thoughts of making separate commands, good? bad? indifferent?

webbedspace commented:
My apologies for my unusually dour tone. At the risk of being unhelpfully repetitive: I believe that there are no advantages to offering `str` commands that just give flat-out incorrect results when querying the length of a string that has *any* non-ASCII characters, and any scripts that require it are equally broken. This is comparably as bad as if, hypothetically, half the commands used 1-indexing for lists. I feel less negative about `split chars` because codepoints are much closer to grapheme clusters, but even the fact that this currently disagrees with `str` about what a char is is a rather serious embarrassment.

Also, I have no idea what `decode grapheme` could possibly mean semantically, and the mere fact that it's being proposed to me indicates an alarming lack of understanding as to what both a grapheme cluster or the `decode` verb actually means ("decode" in Unicode is a byte→string operation, not anything else). Please present some usage examples here.

fdncred commented:
When we were talking about it yesterday, I think we meant that either `decode grapheme` or `into grapheme` would result in the same output, which is a list of graphemes.

webbedspace commented:
How would a list of graphemes help with the `str index-of "ふが"` use-case? You'd have to turn `"ふが"` into `("ふが" | into grapheme)`, and then, I don't even know what list commands would give you the index of a sub-list. This just seems prima facie unworkable.

fdncred reviewed (approved):
I vote for landing this PR and moving on with our lives. We can continue to change and tweak grapheme handling as we go forward. At least this gives users an option to use graphemes if they so choose, even if it's not the default.

WindSoilder commented on crates/nu-command/src/input_handler.rs:
```diff
@@ -86,3 +86,33 @@ where
+
+// A macro for handling the grapheme_cluster related flags.
+#[macro_export]
+macro_rules! grapheme_flags {
```
Hi, what's the reason to define `grapheme_flags` as a macro?  I think it can be handled by rust function.

And since is restricted in str command usage, it'd be better defined inside `strings` module(put it to `strings/mod.rs` is fine), we don't need to put it into `input_handler.rs` file

kubouch commented:
OK, I think it's good. The main points were macro and config which are addressed.

What we meant with the `decode graphemes` is basically `split chars -g` but also accepting binary input, but I think this version is fine. 

We might want to move to graphemes by default, but it might require changing our string type from Rust String to a grapheme string rather than just changing a few commands. There might be places we secretly rely on the current default string behavior.

webbedspace commented:
>We might want to move to graphemes by default, but it might require changing our string type from Rust String to a grapheme string rather than just changing a few commands.

There's not really any such thing as a "grapheme string" as a data structure separate from "native" strings. The way to use graphemes is to `use unicode_segmentation::UnicodeSegmentation` and then call `.graphemes(true)` from Rust Strings (and/or import [unicode_reader](https://docs.rs/unicode_reader/latest/unicode_reader/index.html#)).

kubouch commented:
Yes, that's what I meant. One way to implement it would be to create a grapheme string type that wraps the Rust String and has methods that work on graphemes by default (and can be asked to return the underlying Rust String). Then we'd use that in our Value::String. Something like that would ensure we cover all cases.

fdncred commented:
@webbedspace, if you have time, it would be great to add this our upcoming release blog https://github.com/nushell/nushell.github.io/pull/746 so everyone who reads it will know about your new grapheme work.

eugenesvk commented:
Thanks for this PR, just drooped by after reading the release notes to say that proper human chars, i.e., the grapheme clusters, should be the default, and the other lower-level unintuitive options should require more effort and flags instead of the other way around

bgeron commented:
Also dropping in after reading release notes.

Something to keep in mind: iiuc the clustering rules change with every unicode version. So if we make `-g` the default, then the default behavior of `str length` will automatically change between builds / versions, in a way that's compatible with very few other tools. 

My gut feeling is we should make the distinction clearer and separate the commands — `str byte length` vs. `str char length` etc. Otherwise some group of people is guaranteed to get caught in unexpected ways.

webbedspace commented:
>iiuc the clustering rules change with every unicode version

This is true insofar as each version adds new characters and unused code points are assigned meanings (usually to emoji), so any arbitrary characters in a string are not guaranteed to retain their semantic meaning through time. That being said, if anyone needs the byte length, then they should use `into binary | bytes length`, or make an alias for it.
